rasa/engine/training/graph_trainer.py

```diff
@@ -0,0 +1,161 @@
+
+    def train(
+        self,
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        domain_path: Path,
+        output_filename: Path,
+    ) -> GraphRunner:
+        """Trains and packages a model then returns the prediction graph runner.
+
+        Args:
+            train_schema: The train graph schema.
+            predict_schema: The predict graph schema.
+            domain_path: The path to the domain file.
+            output_filename: The location to save the packaged model.
+
+        Returns:
+            A graph runner loaded with the predict schema.
+
+        """
+        logger.debug("Starting training.")
+
+        pruned_training_schema = self._fingerprint_and_prune(train_schema)
+
+        hooks = [TrainingHook(cache=self._cache, model_storage=self._model_storage)]
+
+        graph_runner = self._graph_runner_class.create(
+            graph_schema=pruned_training_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=pruned_training_schema),
+            hooks=hooks,
+        )
+
+        logger.debug("Train run.")
+
+        graph_runner.run()
+
+        domain = Domain.from_path(domain_path)
+        model_metadata = self._model_storage.create_model_package(
+            output_filename, train_schema, predict_schema, domain
+        )
+
+        return self._graph_runner_class.create(
+            graph_schema=predict_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(
+                graph_schema=predict_schema, model_id=model_metadata.model_id
+            ),
+        )
+
+    def _fingerprint_and_prune(self, train_schema: GraphSchema) -> GraphSchema:
+        fingerprint_schema = self._create_fingerprint_schema(train_schema)
+        fingerprint_graph_runner = self._graph_runner_class.create(
+            graph_schema=fingerprint_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=fingerprint_schema),
+        )
+        logger.debug("Fingerprint run.")
+        fingerprint_run_outputs = fingerprint_graph_runner.run()
+        pruned_training_schema = self._prune_schema(
+            train_schema, fingerprint_run_outputs
+        )
+        return pruned_training_schema
+
+    def _create_fingerprint_schema(self, train_schema: GraphSchema) -> GraphSchema:
+        fingerprint_schema = copy.deepcopy(train_schema)
+        for node_name, schema_node in fingerprint_schema.nodes.items():
+            schema_node.is_target = True
+            if not schema_node.is_input:
+                FingerprintComponent.replace_schema_node(
+                    schema_node, self._cache, node_name
+                )
+        return fingerprint_schema
+
+    def _prune_schema(
+        self,
+        schema: GraphSchema,
+        fingerprint_run_outputs: Dict[Text, Union[FingerprintStatus, Any]],
+    ) -> GraphSchema:
```

Comment:
although this not public I'd add a docstring which explains what we are doing (maybe even give a short example).

Reply:
I will add a small docstring here, and a bigger one to `_fingerprint_and_prune`.